python/demo/demo_pml.py

```diff
@@ -0,0 +1,225 @@
+l_pml = 0.6 * um
+
+# The smaller the mesh_factor, the finer is the mesh
+mesh_factor = 1
+
+# Mesh size inside the wire
+in_wire_size = mesh_factor * 6 * nm
+
+# Mesh size at the boundary of the wire
+on_wire_size = mesh_factor * 3 * nm
+
+# Mesh size in the background
+bkg_size = mesh_factor * 20 * nm
+
+# Mesh size at the boundary
+pml_size = mesh_factor * 20 * nm
+
+# Tags for the subdomains
+au_tag = 1
+bkg_tag = 2
+pml_tag = 3
+# -
+
+model = generate_mesh_wire(
+    radius_wire, l_dom, l_pml, in_wire_size, on_wire_size, bkg_size,
+    pml_size, au_tag, bkg_tag, pml_tag)
+
+mesh, cell_tags, facet_tags = model_to_mesh(
+    model, MPI.COMM_WORLD, 0, gdim=2)
+
+gmsh.finalize()
+MPI.COMM_WORLD.barrier()
+# -
+
+wl0 = 0.4 * um  # Wavelength of the background field
+n_bkg = 1  # Background refractive index
+eps_bkg = n_bkg**2  # Background relative permittivity
+k0 = 2 * np.pi / wl0  # Wavevector of the background field
+deg = np.pi / 180
+theta = 0 * deg  # Angle of incidence of the background field
+
+degree = 3
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), degree)
+V = fem.FunctionSpace(mesh, curl_el)
+
+f = BackgroundElectricField(theta, n_bkg, k0)
+Eb = fem.Function(V)
+Eb.interpolate(f.eval)
+
+x = ufl.SpatialCoordinate(mesh)
+
+alpha = 1
+r_pml = pml_coordinates(x, alpha, k0, l_dom, l_pml)
+J_pml = grad(r_pml)
+
+# Transform the 2x2 Jacobian into a 3x3 matrix.
+J_pml = as_matrix(((J_pml[0, 0], 0          , 0),
+                   (0          , J_pml[1, 1], 0),
+                   (0          , 0          , 1)))
+
+A_pml = inv(J_pml)
+pml_matrix = det(J_pml) * A_pml * transpose(A_pml)
+eps_pml = eps_bkg * pml_matrix
+mu_pml = inv(pml_matrix)
+
+Es = ufl.TrialFunction(V)
+v = ufl.TestFunction(V)
+
+Es_3d = as_vector((Es[0], Es[1], 0))
+v_3d = as_vector((v[0], v[1], 0))
+
+dx = ufl.Measure("dx", mesh, subdomain_data=cell_tags)
+dAu = dx(au_tag)
+dBkg = dx(bkg_tag)
+dDom = dAu + dBkg
+dPml = dx(pml_tag)
+
+reps_au = -1.0782
+ieps_au = 5.8089
+eps_au = reps_au + ieps_au * 1j
+
+D = fem.FunctionSpace(mesh, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.find(au_tag)
+bkg_cells = cell_tags.find(bkg_tag)
+eps.x.array[au_cells] = np.full_like(
+    au_cells, reps_au + ieps_au * 1j, dtype=np.complex128)
+eps.x.array[bkg_cells] = np.full_like(bkg_cells, eps_bkg, dtype=np.complex128)
+eps.x.scatter_forward()
+
+F = - inner(curl_2d(Es), curl_2d(v)) * dDom \
+    + eps * k0 ** 2 * inner(Es, v) * dDom \
+    + k0 ** 2 * (eps - eps_bkg) * inner(Eb, v) * dDom \
+    - inner(mu_pml * curl_2d(Es), curl_2d(v)) * dPml \
+    + k0 ** 2 * inner(eps_pml * Es_3d, v_3d) * dPml
+
+a, L = lhs(F), rhs(F)
+
+problem = fem.petsc.LinearProblem(a, L, bcs=[], petsc_options={
+                                  "ksp_type": "preonly", "pc_type": "lu"})
+Esh = problem.solve()
+
+V_dg = fem.VectorFunctionSpace(mesh, ("DG", degree))
+Esh_dg = fem.Function(V_dg)
+Esh_dg.interpolate(Esh)
+
+with VTXWriter(MPI.COMM_WORLD, "Esh.bp", Esh_dg) as f:
```

Comment:
Instead of MPI.COMM_WORLD use mesh.comm whenever you need a communicator relating to a function or a mesh